### PR TITLE
fix(taskframework): ip change with docker restart cause jobs process interrupted and status cannot be terminated

### PR DESCRIPTION
**Pull request by krihy**
#### What type of PR is this?

type-bug
module-taskframework
#### What this PR does / why we need it:
1.ip change with docker restart cause jobs process interrupted and status cannot be terminated
2.fix query remote log exception
3.fix `odc.task-framework.check-running-job-cron-expression` to `0/10 * * * * ?`,  old `* 0/1 * * * ?` is every second to execute job.

#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-server/src/main/resources/data.sql [hunk 1]
```diff
@@ -734,9 +734,9 @@ INSERT INTO config_system_configuration ( `key`, `value`, `description` ) VALUES
...
-INSERT INTO config_system_configuration ( `key`, `value`, `description` ) VALUES( 'odc.task-framework.do-canceling-job-cron-expression', '0/10 * * * * ?', 'check canceling job cron expression, modify value restart to take affect' ) ON DUPLICATE KEY UPDATE `id` = `id`;
-INSERT INTO config_system_configuration ( `key`, `value`, `description` ) VALUES( 'odc.task-framework.destroy-executor-job-cron-expression', '0/10 * * * * ?', 'check terminate job to destroy executor cron expression, modify value restart to take affect' ) ON DUPLICATE KEY UPDATE `id` = `id`;
+INSERT INTO config_system_configuration ( `key`, `value`, `description` ) VALUES( 'odc.task-framework.check-running-job-cron-expression', '0/10 * * * * ?', 'check running job cron expression, modify value restart to take affect' ) ON DUPLICATE KEY UPDATE `id` = `id`;
+INSERT INTO config_system_configuration ( `key`, `value`, `description` ) VALUES( 'odc.task-framework.do-canceling-job-cron-expression', '0/1 * * * * ?', 'check canceling job cron expression, modify value restart to take affect' ) ON DUPLICATE KEY UPDATE `id` = `id`;
```
Comment (krihy) [inline server/odc-server/src/main/resources/data.sql:738]: old value has been changed in `V_4_2_3_16__update_task_framework_system_config`

Comment (yizhouxw) [reply server/odc-server/src/main/resources/data.sql:738]: all task framework related configurations are involved in version 4.2.4, why there exists 4.2.3 related configuration items?

and, we may remove V_4_2_4_12__update_system_config.sql

File: server/odc-service/src/main/java/com/oceanbase/odc/service/logger/LoggerService.java [hunk 2]
```diff
@@ -122,13 +123,17 @@ public String getLogByTaskFramework(OdcTaskLogLevel level, Long jobId) throws IO
             // process mode when executor is not current host, forward to target
             if (!jobDispatchChecker.isExecutorOnThisMachine(jobEntity)) {
                 ExecutorIdentifier ei = ExecutorIdentifierParser.parser(jobEntity.getExecutorIdentifier());
-                DispatchResponse response = requestDispatcher.forward(ei.getHost(), ei.getPort());
-                return response.getContentByType(new TypeReference<SuccessResponse<String>>() {}).getData();
+                boolean connectable = HttpUtil.isConnectable(ei.getHost(), ei.getPort(), 3);
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/logger/LoggerService.java:126]: magic number used

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 3]
```diff
@@ -54,28 +59,55 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
 
+    private void handleNotLocalAndProcessIsNotExitsJobs(int size, int heartTimeoutPeriod) {
+        // docker restart then ip has been changed and process been destroyed,
+        // so we should find heart timeout not running in local when process mode
+        Page<JobEntity> jobs = getConfiguration().getTaskFrameworkService()
+                .findHeartTimeTimeoutNoLocalJobs(heartTimeoutPeriod, 0, size);
+        jobs.stream().filter(a -> {
+            ExecutorIdentifier ei = ExecutorIdentifierParser.parser(a.getExecutorIdentifier());
+            return !(HttpUtil.isConnectable(ei.getHost(), ei.getPort(), 3));
+        }).forEach(j -> handleJobRetryingOrFailed(j, a -> {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:90]: throw exception in forEach iterator may leads unexpected behavior

Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:91]: ok, add try catch in `handleJobRetryingOrFailed`  to handle exception

File: server/odc-service/src/main/java/com/oceanbase/odc/service/logger/LoggerService.java (hunk 2 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/logger/LoggerService.java]: okey, add properties`getCheckOdcServerCanBeConnectedTimes` in `TaskFrameworkProperties`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 4]
```diff
@@ -54,28 +59,59 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
+    private void handleJobRetryingOrFailed(JobEntity a, Consumer<JobIdentity> jobIdentityConsumer) {
         try {
-            getConfiguration().getJobDispatcher().destroy(JobIdentity.of(a.getId()));
-        } catch (JobException e) {
-            throw new TaskRuntimeException(e);
+            getConfiguration().getTransactionManager().doInTransactionWithoutResult(() -> {
+                doHandleJobRetryingOrFailed(a, jobIdentityConsumer);
+            });
+        } catch (Exception e) {
+            log.warn("handleJobRetryingOrFailed occur exception:", e);
```
[the remark this answers is not in the document]
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java]: ok,  `SilentExecutor` has been added like below

**Review by yizhouxw — DISMISSED** [review]
lgtm

File: server/odc-common/src/main/java/com/oceanbase/odc/common/util/SilentExecutor.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @author yaobin
+ * @date 2024-03-25
+ * @since 4.2.4
+ */
+@Slf4j
+public class SilentExecutor {
```
Comment (yhilmare) [inline server/odc-common/src/main/java/com/oceanbase/odc/common/util/SilentExecutor.java:26]: why should we need this?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 6]
```diff
@@ -54,51 +62,86 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
-
-    private void handleJobRetryingOrCanceled(JobEntity a) {
-        getConfiguration().getTransactionManager().doInTransactionWithoutResult(() -> {
-            doHandleJobRetryingOrFailed(a);
-        });
-
+        for (JobEntity j : jobs) {
+            SilentExecutor.executeSafely("handleJobRetryingOrFailed", () -> {
+                handleJobRetryingOrFailed(j, a -> {
+                    try {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:78]: why should we catch the exception here?
doesn't you catch this exception in SilentExecutor?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 7]
```diff
@@ -54,51 +62,86 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
         TaskFrameworkProperties taskFrameworkProperties = getConfiguration().getTaskFrameworkProperties();
         int size = taskFrameworkProperties.getSingleFetchCheckHeartTimeoutJobRows();
         int heartTimeoutPeriod = taskFrameworkProperties.getJobHeartTimeoutSeconds();
+        handleGeneralHeartTimeoutJobs(size, heartTimeoutPeriod);
+        if (taskFrameworkProperties.getRunMode().isProcess()) {
+            handleNotLocalAndProcessIsNotExitsJobs(size, heartTimeoutPeriod);
+        }
+    }
+
+    private void handleGeneralHeartTimeoutJobs(int size, int heartTimeoutPeriod) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:71]: what's the unit of this param? `heartTimeoutPeriod`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 8]
```diff
@@ -54,51 +62,86 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
+    }
 
-        if (checkJobIfRetryNecessary(a)) {
-            log.info("Need to restart job, destroy old executor completed, jobId={}.", a.getId());
-            int rows = getConfiguration().getTaskFrameworkService()
-                    .updateStatusDescriptionByIdOldStatusAndExecutorDestroyed(a.getId(), JobStatus.RUNNING,
-                            JobStatus.RETRYING, "Heart timeout and retrying job");
-            if (rows > 0) {
-                log.info("Job {} set status to RETRYING.", a.getId());
+    private void handleJobRetryingOrFailed(JobEntity a, Consumer<JobIdentity> jobIdentityConsumer) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:116]: why should we need this consumer? does this consumer has another implementation ?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/service/StdTaskFrameworkService.java [hunk 9]
```diff
@@ -155,6 +155,23 @@ public Page<JobEntity> findHeartTimeTimeoutJobs(int timeoutSeconds, int page, in
...
+                .and((root, query, cb) -> getHeartTimeoutPredicate(root, cb, timeoutSeconds))
+                .and(getNotLocalExecutor());
+        return page(condition, page, size);
+    }
+
+    private static Specification<JobEntity> getNotLocalExecutor() {
+        return (root, query, cb) -> cb.and(
+                cb.equal(root.get(JobEntityColumn.RUN_MODE), TaskRunMode.PROCESS),
+                cb.or(cb.notLike(root.get(JobEntityColumn.EXECUTOR_IDENTIFIER),
+                        "%" + StringUtils.escapeLike(SystemUtils.getLocalIpAddress()) + "%"),
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/service/StdTaskFrameworkService.java:171]: why can we identify a machine by its ip?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 10]
```diff
@@ -54,51 +62,86 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
-        } catch (JobException e) {
-            throw new TaskRuntimeException(e);
+    private void handleNotLocalAndProcessIsNotExitsJobs(int size, int heartTimeoutPeriod) {
+        // ip has been changed docker restart and process has been interrupted,
+        // so we should find jobs which heart timeout and not running in local
+        Page<JobEntity> jobs = getConfiguration().getTaskFrameworkService()
+                .findHeartTimeTimeoutNotLocalJobs(heartTimeoutPeriod, 0, size);
+
+        List<JobEntity> jobList = jobs.stream().filter(a -> {
+            ExecutorIdentifier ei = ExecutorIdentifierParser.parser(a.getExecutorIdentifier());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:96]: what if executor identifier is null?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 11]
```diff
@@ -54,51 +62,86 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
+        }).collect(Collectors.toList());
+
+        for (JobEntity j : jobList) {
+            SilentExecutor.executeSafely("handleJobRetryingOrFailed", () -> {
+                handleJobRetryingOrFailed(j, a -> {
+                    int rows = getConfiguration().getTaskFrameworkService().updateExecutorToDestroyed(a.getId());
+                    if (rows > 0) {
+                        log.info("Executor is not exists, update job executor to destroyed, jobId={}", a.getId());
+                    } else {
+                        throw new TaskRuntimeException("update executor to destroyed failed, jobId=" + a.getId());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:108]: who will catch this exception?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java [hunk 12]
```diff
@@ -92,4 +99,40 @@ public static <T> T request(String method, String uri, Map<String, String> heade
...
+                break;
+            }
+        }
+        return connectable;
+    }
+
+    public static boolean isConnectable(String server, int servPort) {
+        try {
+            InetAddress ad = InetAddress.getByName(server);
+            boolean state = ad.isReachable(1000);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java:116]: is 1 second too short?

Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java:116]: what if icmp protocol is disable?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java [hunk 13]
```diff
@@ -92,4 +99,40 @@ public static <T> T request(String method, String uri, Map<String, String> heade
...
+        } catch (IOException e) {
+            log.warn("Test target server can not be connected, servIp={}, servPort={}", server, servPort);
+            return false;
+        }
+        try (Socket socket = new Socket()) {
+            socket.setReceiveBufferSize(8192);
+            socket.setSoTimeout(1000);
+            SocketAddress address = new InetSocketAddress(server, servPort);
+            // Test tcp can be connected succeed
+            socket.connect(address, 1000);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java:130]: when shall we close this socket?

File: server/odc-common/src/main/java/com/oceanbase/odc/common/util/SilentExecutor.java (hunk 5 above)
Comment (krihy) [reply server/odc-common/src/main/java/com/oceanbase/odc/common/util/SilentExecutor.java:26]: see above comment, to handle exception in for loop

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 7 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:71]: okey, heartTimeoutPeriod  -> heartTimeoutSeconds

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 6 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:78]: JobException is Check Exception, so i change it to RuntimeException, so no need to throw exception

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 8 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:116]: yes , `handleGeneralHeartTimeoutJobs` and `handleNotLocalAndProcessIsNotExitsJobs` has two implementation

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 10 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:96]: executor identifier will not be null if job is `RUNNING`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 11 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:108]: exception will be catch in `getConfiguration().getTransactionManager().doInTransactionWithoutResult` and transaction will be rollback

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 14]
```diff
@@ -54,51 +62,79 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
+                                JobStatus.RETRYING, "Heart timeout and retrying job");
+                if (rows > 0) {
+                    log.info("Job {} set status to RETRYING.", a.getId());
+                }
+
+            } else {
+                log.info("No need to restart job, try to set status to FAILED, jobId={}.", a.getId());
+                TaskFrameworkProperties taskFrameworkProperties = getConfiguration().getTaskFrameworkProperties();
+                int rows = getConfiguration().getTaskFrameworkService()
+                        .updateStatusToCanceledWhenHeartTimeout(a.getId(),
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:128]: updateToFailed

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 15]
```diff
@@ -54,51 +62,79 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
-            throw new TaskRuntimeException(e);
+    private void handleNotLocalAndProcessIsNotExitsJobs(int size, int heartTimeoutSeconds) {
+        // ip has been changed docker restart and process has been interrupted,
+        // so we should find jobs which heart timeout and not running in local
+        Page<JobEntity> jobs = getConfiguration().getTaskFrameworkService()
+                .findHeartTimeTimeoutNotLocalJobs(heartTimeoutSeconds, 0, size);
+
+        List<JobEntity> jobList = jobs.stream().filter(a -> {
+            ExecutorIdentifier ei = ExecutorIdentifierParser.parser(a.getExecutorIdentifier());
+            return !(HttpUtil.isConnectable(ei.getHost(), ei.getPort(),
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:92]: port can not be a constant

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java [hunk 16]
```diff
@@ -54,51 +62,79 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
         int size = taskFrameworkProperties.getSingleFetchCheckHeartTimeoutJobRows();
         int heartTimeoutPeriod = taskFrameworkProperties.getJobHeartTimeoutSeconds();
-        // find heart timeout job
-        Page<JobEntity> jobs = getConfiguration().getTaskFrameworkService()
-                .findHeartTimeTimeoutJobs(heartTimeoutPeriod, 0, size);
-        jobs.forEach(this::handleJobRetryingOrCanceled);
-
+        handleGeneralHeartTimeoutJobs(size, heartTimeoutPeriod);
+        if (taskFrameworkProperties.getRunMode().isProcess()) {
+            handleNotLocalAndProcessIsNotExitsJobs(size, heartTimeoutPeriod);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java:67]: dup code~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 6 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java]: done, fetch 10000 timeout in process mode and destroy process in BaseJobCaller

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 10 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java]: remove this code

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java (hunk 12 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/HttpUtil.java]: add odc server healthy check in HttpUtil

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 14 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 15 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java (hunk 16 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java]: done

**Review by yhilmare — APPROVED** [review]
lgtm

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**